hermeto/core/package_managers/gomod.py

```diff
@@ -881,24 +950,33 @@ def _setup_go_toolchain(go_mod_file: RootedPath) -> Go:
             ),
         )
 
-    if target_version >= GO_121:
-        # Project makes use of Go >=1.21:
-        # - always use the 'X.Y.0' toolchain to make sure GOTOOLCHAIN=auto fetches anything newer
-        # - container environments need to have it pre-installed
-        # - local environments will always install 1.21.0 SDK and then pull any newer toolchain
-        go = Go(release="go1.21.0")
-    elif go_base_version >= GO_121:
-        # Starting with Go 1.21, Go doesn't try to be semantically backwards compatible in that the
-        # 'go X.Y' line now denotes the minimum required version of Go, no a "suggested" version.
-        # What it means in practice is that a Go toolchain >= 1.21 enforces the biggest common
-        # toolchain denominator across all dependencies and so if the input project specifies e.g.
-        # 'go 1.19' and **any** of its dependencies specify 'go 1.21' (or higher), then the default
-        # 1.21 toolchain will bump the input project's go.mod file to make sure the minimum
-        # required Go version is met across all dependencies. That is a problem, because it'll lead
-        # to fatal build failures forcing everyone to update their build recipes. Note that at some
-        # point they'll have to do that anyway, but until majority of projects in the ecosystem
-        # adopt 1.21, we need a fallback to an older toolchain version.
-        go = Go(release="go1.20")
+    # If we cannot find a matching toolchain, we'll try to fallback to a 1.21 one
+    matching_toolchains = filter(lambda t: t.version >= target_version, installed_toolchains)
+    try:
+        go = min(matching_toolchains)
```

Comment:
Why the minimum of the matching ones here?

Reply:
So the filter returns everything `>= <user preferred toolchain>` and so the premise here is to select the closest one matching, so e.g. if a user requires `1.24.8` and we have `1.24.9, 1.25.0, 1.25.1` then we'd go with `1.24.9` as there should be the least number of surprises.